File: scripts/parse-results.R
#!/usr/bin/env Rscript

options(warn=-1)
suppressPackageStartupMessages({
        require(stringr)
})

# Parsing arguments
args = commandArgs(trailingOnly=TRUE)

# reading command arguments
input = args[1]
output = args[2]
region = args[3]

# defining functions
GetExonID = function(x){
    full_name = x[5]
    first_split = str_split(full_name, pattern = "=", simplify = T)
    if(ncol(first_split) == 1){ 
    	geneID = first_split[,1] 
    }
    else{
    	first_split = first_split[,2]
        geneID = str_split(first_split,pattern = ";", simplify = T)[1,1]
    }
}

GetGeneIDForExons = function(x){
    full_gene_info = x[5]
    first_split = str_split(full_gene_info, pattern = "gene=", simplify = T)
    first_split = first_split[,2]
    geneID = str_split(first_split,pattern = ";", simplify = T)[1,1]
    paste0("gene-", geneID)
}

GetUCENames = function(x){
	UCE = x[1]
	paste0(str_split(UCE, pattern="_", simplify=T)[,1], ".nexus")
}

GetGeneIDForIntrons = function(x){
    full_gene_info = x[5]
    first_split = str_split(full_gene_info, pattern = "ID=", simplify = T)
    # make another split by ";" to get rid of additional information
    # and get only the gene ID
    first_split = first_split[,2]
    str_split(first_split,pattern = ";", simplify = T)[1,1]
}

# read dataframe
df = read.table(input, header=F, sep="\t")

if(region == "exon"){
	df$UCE = apply(df, MARGIN = 1, FUN = GetUCENames)
	df$Exon = apply(df, MARGIN = 1, FUN = GetExonID)
	df$Gene = apply(df, MARGIN = 1, FUN = GetGeneIDForExons)
	df = df[,c(6,7,8)]
}
if(region == "intron"){
	df$UCE = apply(df, MARGIN = 1, FUN = GetUCENames)
	df$Intron = apply(df, MARGIN=1, FUN= GetGeneIDForIntrons)
	df = df[,c(6,7)]
}
if(region == "intergenic"){
	df$UCE = apply(df, MARGIN = 1, FUN = GetUCENames)
	df = df$UCE
}
write.table(df, output, row.names = F, quote=F, col.names = F, sep="\t")

File: scripts/parse-results.py
#!/usr/bin/env python
import sys

def read_df(df_path, filt, output, region):
    # open output and input files
    with open(df_path,'r') as df:
        with open(output, 'w') as outdf:
            for line in df.readlines():
                # input is tab delimited:
                row = line.strip().split('\t')
                # all types will get the UCE name filtered
                UCE_name = row[0]
                UCE_name_filtered = UCE_name.split(filt)[0]+'.nexus'
                # now, each type will have diff. treatments
                if region == "exon":
                    full_ID = row[4]
                    # full ID example for exons:
                    # BASE:GB51190;gbkey=mRNA;gene=LOC411181;product=sodium...
                    # GeneID is between 'gene=' and ';'
                    first_split = full_ID.split('gene=')[1]
                    second_split = first_split.split(';')[0]
                    gene_ID = 'gene-'+second_split
                    # same for exon ID:
                    # ID=exon-XM_016912896.2-10;Parent=rna-XM_0169128...
                    # ExonID is between 'ID=' and ';'
                    first_split = full_ID.split('ID=')[1]
                    exon_ID = first_split.split(';')[0]
                    # all done, let's write it
                    line2write = '\t'.join([UCE_name_filtered, exon_ID, gene_ID])+'\n'
                    outdf.write(line2write)
                elif region == "intron":
                    full_ID = row[4]
                    # full ID example for introns:
                    # ID=gene-LOC410882
                    # Just need to remove 'ID='
                    gene_ID = full_ID.split('ID=')[1]
                    # all done, let's write it
                    line2write = '\t'.join([UCE_name_filtered, gene_ID])+'\n'
                    outdf.write(line2write)
                elif region == 'intergenic':
                    # nothing to do, just write UCE name filtered
                    outdf.write(UCE_name_filtered+'\n')

def main():
    # read arguments
    args = sys.argv
    df_path = args[1]
    filt = args[2]
    output = args[3]
    region = args[4]
    print('df: {}'. format(df_path))
    print('filt: {}'. format(filt))
    print('output: {}'. format(output))
    print('region: {}'. format(region))
    # call function
    read_df(df_path, filt, output, region)
    
if __name__ == '__main__':
    main()